test/tstNeighborList.cpp

```diff
@@ -0,0 +1,249 @@
+/****************************************************************************
+ * Copyright (c) 2017-2023 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+// clang-format off
+#include "boost_ext/KokkosPairComparison.hpp"
+#include "boost_ext/TupleComparison.hpp"
+#include "boost_ext/CompressedStorageComparison.hpp"
+// clang-format on
+
+#include "ArborXTest_StdVectorToKokkosView.hpp"
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsExpandHalfToFull.hpp>
+#include <ArborX_NeighborList.hpp>
+
+#include <Kokkos_Random.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+namespace Test
+{
+using ArborXTest::toView;
+
+template <class ExecutionSpace>
+auto make_random_cloud(ExecutionSpace const &space, int n)
+{
+  Kokkos::View<ArborX::Point *, ExecutionSpace> points(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing, "Test::points"),
+      n);
+  using RandomPool = Kokkos::Random_XorShift64_Pool<ExecutionSpace>;
+  RandomPool random_pool(5374857);
+
+  Kokkos::parallel_for(
+      "Test::generate_random_points",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int i) {
+        typename RandomPool::generator_type generator = random_pool.get_state();
+        auto const x = generator.frand(0.f, 1.f);
+        auto const y = generator.frand(0.f, 1.f);
+        auto const z = generator.frand(0.f, 1.f);
+        points(i) = {x, y, z};
+        random_pool.free_state(generator);
+      });
+
+  return points;
+}
+
+struct Filter
+{
+  template <class Predicate, class OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &predicate, int i,
+                                  OutputFunctor const &out) const
+  {
+    int const j = getData(predicate);
+    if (i < j)
+    {
+      out(i);
+    }
+  }
+};
+
+template <class Points>
+struct RadiusSearch
+{
+  Points points;
+  float radius;
+};
```

Comment:
How about we provide we do a PR after this one where we provide this utility (predicates based on points and a radius) and use it everywhere? It's been appearing all over the place.